### DISCUSSION TITLE
child machine magically restarted

### DISCUSSION THREAD

**Discussion by augustblack** [statelyai/xstate#3325] [General]
have a look at:

https://codesandbox.io/s/bold-glade-s7z891

first click 'add player' to spawn a child player machine.  

then, if you click the buttons to set the 'term' and 'kind' on the child playlist (of the child player machine), it will set the playlist.context.term and kind respectively (and you will see the term and kind change from 'default term' to 'new term').  

However, if you then click "send pl add"  it will send the event PLAYLIST_ADD to the player machine and will kill the old playlist child machine and cause a new playlist child machine to start.

**Comment by Andarist** [comment]
Please always try to list concise repro steps and clear descriptions of the actual & expected results (how I can observe the expectation mismatch etc).

If I understood correctly your question then you probably want to use internal transitions there for the `PLAYLIST_SET` event. Please check if this matches your expectations: https://codesandbox.io/s/pedantic-zhukovsky-pz9ed7

**Comment by augustblack** [comment]
Hi @Andarist  

Thanks for taking a look.

As I understand it, if you invoke a child machine at the root of a parent machine, the child machine should run and **not restart** until it either reaches a final state or the parent ends, correct?

In my example, the player.machine invokes the playlist.machine as a child.  However, when running the PLAYLIST_SET event the first time, the playlist child machine is restarted.   Should that be happening? 

Or, is that what an internal transition is supposed to guard against?

The docs say "...an internal transition is a transition that transitions to a descendant target state without exiting the source state."  
I would think all the states in my example (startup, loading, empty) are sibling states, no?   Is transitioning from "startup" to "loading" supposed to be internal?

**Comment by augustblack** [comment]
Am I understanding this correctly that this is a bug?  That a child machine invoked at the root level should not be restarted?

**Reply by Andarist** [reply]
The root itself is just another (anonymous) state. So currently it doesn't get any special treatment and you can "reenter" it just like any other state. We plan to change this in the next major version though.
